File: frontend/src/lib/indicators.ts
/**
 * API client for indicator endpoints.
 *
 * Follows the same apiFetch pattern used in lib/api.ts.
 * All functions call /api/indicators/... endpoints served by the backend.
 */

const API_BASE =
  process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000";

// ---------- Types ----------

export interface IndicatorDataPoint {
  date: string;
  value: number;
}

export interface IndicatorData {
  slug: string;
  name: string;
  units: string;
  methodology: string;
  frequency: string;
  data: IndicatorDataPoint[];
  latest_value: number | null;
  latest_date: string | null;
}

export interface IndicatorSummary {
  slug: string;
  name: string;
  units: string;
  latest_value: number | null;
  latest_date: string | null;
  trend: string | null;
  sparkline: number[];
}

export interface IndicatorOverview {
  methodology: string;
  indicators: IndicatorSummary[];
}

export interface NIPAMappingDoc {
  marx_category: string;
  nipa_table: string;
  nipa_line: number;
  nipa_description: string;
  operation: string;
  notes: string;
}

export interface IndicatorMethodologyDoc {
  indicator_slug: string;
  indicator_name: string;
  formula_latex: string;
  interpretation: string;
  mappings: NIPAMappingDoc[];
  citations: string[];
}

export interface MethodologyDocResponse {
  methodology_slug: string;
  methodology_name: string;
  indicators: IndicatorMethodologyDoc[];
}

// ---------- Internal fetch helper ----------

async function indicatorFetch<T>(
  path: string,
  params?: Record<string, string | undefined>
): Promise<T> {
  const url = new URL(path, API_BASE);
  if (params) {
    for (const [key, value] of Object.entries(params)) {
      if (value !== undefined && value !== "") {
        url.searchParams.set(key, value);
      }
    }
  }

  const res = await fetch(url.toString(), {
    headers: { Accept: "application/json" },
  });

  if (!res.ok) {
    throw new Error(`API error ${res.status}: ${res.statusText}`);
  }

  return res.json() as Promise<T>;
}

// ---------- Public API ----------

/**
 * Fetch overview of all indicators with latest values and sparkline data.
 */
export async function fetchIndicatorOverview(
  methodology?: string
): Promise<IndicatorOverview> {
  return indicatorFetch<IndicatorOverview>("/api/indicators/", {
    methodology,
  });
}

/**
 * Fetch time-series data for a single indicator.
 */
export async function fetchIndicatorData(
  slug: string,
  params?: { methodology?: string; start?: string; end?: string }
): Promise<IndicatorData> {
  return indicatorFetch<IndicatorData>(
    `/api/indicators/${encodeURIComponent(slug)}`,
    params as Record<string, string | undefined>
  );
}

/**
 * Fetch the same indicator computed under all methodologies for comparison.
 */
export async function fetchIndicatorComparison(
  slug: string
): Promise<{ methodologies: Record<string, IndicatorData> }> {
  return indicatorFetch<{ methodologies: Record<string, IndicatorData> }>(
    `/api/indicators/${encodeURIComponent(slug)}/compare`
  );
}

/**
 * Fetch methodology documentation (formulas, NIPA mappings, citations).
 * If methodologySlug is provided, returns docs for that methodology only.
 */
export async function fetchMethodologyDocs(
  methodologySlug?: string
): Promise<MethodologyDocResponse> {
  const path = methodologySlug
    ? `/api/indicators/methodology/${encodeURIComponent(methodologySlug)}`
    : "/api/indicators/methodology";
  return indicatorFetch<MethodologyDocResponse>(path);
}

File: backend/ecpm/schemas/indicators.py
"""Pydantic v2 response models for indicator API endpoints.

Defines the API contract for indicator data, overview summaries,
and methodology documentation endpoints.
"""

from __future__ import annotations

import datetime as dt
from typing import Optional

from pydantic import BaseModel, ConfigDict


class IndicatorDataPoint(BaseModel):
    """A single data point in an indicator time-series."""

    model_config = ConfigDict(from_attributes=True)

    date: dt.datetime
    value: float


class IndicatorResponse(BaseModel):
    """Full time-series response for a single indicator."""

    model_config = ConfigDict(from_attributes=True)

    slug: str
    name: str
    units: str
    methodology: str
    frequency: str
    data: list[IndicatorDataPoint]
    latest_value: Optional[float] = None
    latest_date: Optional[dt.datetime] = None


class IndicatorSummary(BaseModel):
    """Summary card data for one indicator on the overview page."""

    model_config = ConfigDict(from_attributes=True)

    slug: str
    name: str
    units: str
    latest_value: Optional[float] = None
    latest_date: Optional[dt.datetime] = None
    trend: Optional[str] = None  # "rising", "falling", "flat"
    sparkline: list[float] = []  # last N values for overview card


class IndicatorOverviewResponse(BaseModel):
    """Overview page response: all indicator summaries under one methodology."""

    model_config = ConfigDict(from_attributes=True)

    methodology: str
    indicators: list[IndicatorSummary]


class NIPAMappingDoc(BaseModel):
    """Documentation for one NIPA-to-Marx mapping."""

    model_config = ConfigDict(from_attributes=True)

    marx_category: str
    nipa_table: str
    nipa_line: int
    nipa_description: str
    operation: str
    notes: str = ""


class IndicatorMethodologyDoc(BaseModel):
    """Documentation for one indicator under a specific methodology."""

    model_config = ConfigDict(from_attributes=True)

    indicator_slug: str
    indicator_name: str
    formula_latex: str
    interpretation: str
    mappings: list[NIPAMappingDoc]
    citations: list[str]


class MethodologyDocResponse(BaseModel):
    """Full methodology documentation: all indicators under one methodology."""

    model_config = ConfigDict(from_attributes=True)

    methodology_slug: str
    methodology_name: str
    indicators: list[IndicatorMethodologyDoc]
